src/utils/manifold_features.jl

```diff
@@ -0,0 +1,177 @@
+        try
+            f(M, p)
+            push!(features, f)
+        catch
+        end
+    end
+    # c) signature f(M, p, X)
+    for f in [exp, geodesic, is_vector, norm, retract]
+        try
+            f(M, p, X)
+            push!(features, f)
+        catch
+        end
+    end
+    # d) signatures f(M, p, q)
+    for f in [distance, inverse_retract, log, shortest_geodesic]
+        try
+            f(M, p, p)
+            push!(features, f)
+        catch
+        end
+    end
+    # e) signature f(M, p, X, q)
+    for f in [parallel_transport_to, vector_transport_to]
+        try
+            f(M, p, X, p)
+            push!(features, f)
+        catch
+        end
+    end
+    # f) signature f(M, p, X, Y)
+    for f in [parallel_transport_direction, vector_transport_direction]
+        try
+            f(M, p, X, X)
+            push!(features, f)
+        catch
+        end
+    end
+    # f) signature f(M, p, X, Y, Z)
+    for f in [riemann_tensor]
+        try
+            f(M, p, X, X, X)
+            push!(features, f)
+        catch
+        end
+    end
+    return features
+end
+
+"""
+    ManifoldFeatures
+
+Collect a set of features available on a manifold.
+
+# Fields
+
+* `functions`
+* `retractions`
+* `inverseretractions`
+* `vector_transports`
+* `tolerances` – based on functions or symbols provide a tolerance
+"""
+struct ManifoldFeatures{F<:Function}
+    functions::Vector{F}
+    retractions::Vector{AbstractRetractionMethod}
+    inverse_retractions::Vector{AbstractInverseRetractionMethod}
+    vector_transports::Vector{AbstractVectorTransportMethod}
+end
+function ManifoldFeatures(;
+    functions::Vector{F}=Functions[],
+    retractions=Vector{AbstractRetractionMethod}[],
+    inverse_retractions=Vector{AbstractInverseRetractionMethod}[],
+    vector_transports=Vector{AbstractVectorTransportMethod}[],
+) where {F<:Function}
+    return ManifoldFeatures{F}(
+        functions,
+        retractions,
+        inverse_retractions,
+        vector_transports,
+    )
+end
+function ManifoldFeatures(
+    M::AbstractManifold;
+    functions::Vector{F}=find_manifold_functions(M),
+    retractions=Vector{AbstractRetractionMethod}[],
+    inverse_retractions=Vector{AbstractInverseRetractionMethod}[],
+    vector_transports=Vector{AbstractVectorTransportMethod}[],
+) where {F<:Function}
+    return ManifoldFeatures{F}(
+        functions,
+        retractions,
+        inverse_retractions,
+        vector_transports,
+    )
+end
+
+@doc """
+    ManifoldExpectations
+
+For the features from [`ManifoldFeatures`](@ref) this struct provides expected
+values and tolerances
+
+* `values`     a dictionary `:Symbol->value` for expected values, e.g. `:dimension`
+* `strings`    a dictionary `:Symbol->String` for expected strings, e.g. `:repr` for `repr(M)`
+* `tolerances` a dictionary `:Symbol->Float64` for tolerances in checks of functions.
+"""
+struct ManifoldExpectations
+    values::Dict{Symbol,Int}
```

Comment:
Maybe not necessarily `Int`?

Reply:
I was a bit unsure here, but yeah, in principle even vectors would be fine, since it should just collect values we expect – I also was not so sure whether splitting of the strings is necessary, though then the dict is `Any` so that might slow down things – or we maybe also do not care so much.

Reply:
I think `Any` here would be perfectly fine. After all, it's code that would be used very few times so there is not much point in making it type-stable.

Reply:
Ok, i will check whether then the strings can be added here as well.